### PR TITLE
Merge healthcheck and consensus endpoints [ECR-1665]

### PR DESCRIPTION
**Pull request by SergeiMal**

### CODE REVIEW

File: exonum/src/api/public/system.rs [hunk 1]
```diff
@@ -67,15 +89,15 @@ impl SystemApi {
...
-    fn healthcheck_info(self, router: &mut Router) {
-        let healthcheck = move |_: &mut Request| -> IronResult<Response> {
-            let info = HealthCheckInfo {
-                connectivity: !self.shared_api_state.peers_info().is_empty(),
-            };
-            self.ok_response(&serde_json::to_value(info).unwrap())
-        };
-        router.get("/v1/healthcheck", healthcheck, "healthcheck");
-    }
+    //fn healthcheck_info(self, router: &mut Router) {
```
Comment (vldm) [inline exonum/src/api/public/system.rs]: remove comments

File: exonum/src/api/public/system.rs [hunk 2]
```diff
@@ -85,22 +107,48 @@ impl SystemApi {
...
+
+            let info = ConnectivityInfo {
+                consensus_status,
+                connectivity,
             };
+
             self.ok_response(&serde_json::to_value(info).unwrap())
         };
-        router.get("/v1/consensus_status", consensus_status, "consensus_status");
+        router.get(
```
Comment (vldm) [inline exonum/src/api/public/system.rs]: i am not sure about endpoint naming, maybe healthcheck would be better?

File: testkit/tests/system_api.rs [hunk 3]
```diff
@@ -53,12 +54,3 @@ fn test_network() {
     assert_eq!(info.protocol_version, PROTOCOL_MAJOR_VERSION);
     assert!(info.services.is_empty());
 }
-
-#[test]
-fn test_consensus_status_false() {
```
Comment (vldm) [inline testkit/tests/system_api.rs]: Can we rewrite test using new DTO?

File: exonum/src/api/public/system.rs [hunk 4]
```diff
@@ -25,18 +25,40 @@ struct MemPoolInfo {
     pub size: usize,
 }
 
-#[doc(hidden)]
+/// Contains amount of connected peers
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: 1. Please add dots here and below.
2. If I don't miss anything, we use verbs for methods and nouns for structures in documentation (examples could be found in [node/mod.rs](https://github.com/exonum/exonum/blob/master/exonum/src/node/mod.rs#L67).

File: exonum/src/api/public/system.rs [hunk 5]
```diff
@@ -25,18 +25,40 @@ struct MemPoolInfo {
...
-/// in current state.
+/// Shows connectivity status of the node
 #[derive(Deserialize, Serialize, Clone, Debug, PartialEq)]
-pub struct ConsensusStatusInfo {
-    /// Consensus status: true - if consensus achieved, else - false.
-    pub status: bool,
+pub enum ConnectivityStatus {
+    /// The node has no peers
+    NotConnected,
+    /// The node has amount of peers
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: `Has amount of peers` sounds weird to me.

File: exonum/src/api/public/system.rs [hunk 6]
```diff
@@ -25,18 +25,40 @@ struct MemPoolInfo {
...
+pub enum ConsensusStatus {
+    /// Consensus disabled
+    Disabled,
+    /// Consensus enabled but not active
+    Enabled,
+    /// Consensus enabled and active
+    Active,
+}
+
+/// Information about connectivity and consensus
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: Documentation for this structure and its fields doesn't seem being helpful at all.

File: exonum/src/api/public/system.rs [hunk 7]
```diff
@@ -85,22 +107,48 @@ impl SystemApi {
         router.get("/v1/user_agent", user_agent, "user_agent");
     }
 
-    fn consensus_status_info(self, router: &mut Router) {
-        let consensus_status = move |_: &mut Request| -> IronResult<Response> {
-            let info = ConsensusStatusInfo {
-                status: self.shared_api_state.consensus_status(),
+    fn connectivity_status_info(self, router: &mut Router) {
+        let connectivity_status = move |_: &mut Request| -> IronResult<Response> {
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: `Actix` PR is already merged. Why is this implementation use `Iron`?

File: exonum/src/api/public/system.rs [hunk 8]
```diff
@@ -85,22 +107,48 @@ impl SystemApi {
...
+            let peers_info = self.shared_api_state.peers_info();
+
+            let connectivity = if peers_info.is_empty() {
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: I'd suggest to move this logic to the separate method, e.g.
```
let connectivity = self.get_connectivity_status();
```

File: exonum/src/api/public/system.rs [hunk 9]
```diff
@@ -85,22 +107,48 @@ impl SystemApi {
...
+                ConnectivityStatus::NotConnected
+            } else {
+                ConnectivityStatus::Connected(PeersAmount {
+                    amount: peers_info.len(),
+                })
+            };
+
+            let consensus_status = if self.shared_api_state.is_enabled() {
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: I'd suggest to move this logic to the separate method, e.g.
```
let connectivity = self.get_consensus_status();
```

File: exonum/src/api/public/system.rs [hunk 10]
```diff
@@ -85,22 +107,48 @@ impl SystemApi {
...
+            "connectivity_status",
+        );
     }
 }
 
 impl Api for SystemApi {
     fn wire(&self, router: &mut Router) {
         self.clone().mempool_info(router);
-        self.clone().healthcheck_info(router);
+        //self.clone().healthcheck_info(router);
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: Commented line.

File: exonum/src/api/public/system.rs [hunk 11]
```diff
@@ -85,22 +107,48 @@ impl SystemApi {
         router.get("/v1/user_agent", user_agent, "user_agent");
     }
 
-    fn consensus_status_info(self, router: &mut Router) {
-        let consensus_status = move |_: &mut Request| -> IronResult<Response> {
-            let info = ConsensusStatusInfo {
-                status: self.shared_api_state.consensus_status(),
+    fn connectivity_status_info(self, router: &mut Router) {
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: This method should be renamed, it's not only about connectivity.

File: testkit/tests/system_api.rs [hunk 12]
```diff
@@ -18,18 +18,19 @@ extern crate exonum_testkit;
...
-                   public::{ConsensusStatusInfo, HealthCheckInfo}},
+                   public::{ConnectivityInfo, ConnectivityStatus, ConsensusStatus}},
              helpers::user_agent,
              messages::PROTOCOL_MAJOR_VERSION};
 use exonum_testkit::{ApiKind, TestKitBuilder};
 
 #[test]
-fn test_healthcheck_connectivity_false() {
-    let testkit = TestKitBuilder::validator().with_validators(2).create();
+fn test_connectivity_status_false() {
```
Comment (popzxc) [inline testkit/tests/system_api.rs]: I'd expect more tests for this endpoint with different combinations tested (ConsensusEnabled/Connected, ConsensusDisabled/Connected, etc).

Comment (SergeiMal) [reply testkit/tests/system_api.rs]: As I can see, this tests able to check whether the endpoint return expected result and correctness of serialisation. To perform more complex test need to use sandbox, and work on it in context of another task, I think.

Comment (popzxc) [reply testkit/tests/system_api.rs]: Sure.

File: exonum/src/api/public/system.rs [hunk 13]
```diff
@@ -0,0 +1,148 @@
+// Copyright 2018 The Exonum Team
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: Seems like this file is some merge error. New hierarchy doesn't contain `/src/api/public` folder.

File: exonum/src/api/public/system.rs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+// limitations under the License.
+
+//! Public system API.
+
+use api::{ServiceApiScope, ServiceApiState};
+use blockchain::{Schema, SharedNodeState};
+use helpers::user_agent;
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+/// Information about the current state of the Node memory pool.
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: Offtop: BTW please move this doc line above `#[derive(..)]` statement.

**Comment by popzxc** [comment]
I'd suggest to run all of the following:
-tests
-rustfmt
-clippy
-cspell

before pushing. It'd help to detect a lot of problems even before the review.

File: exonum/src/api/public/system.rs [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    Enabled,
+    /// Consensus enabled and the node has enough connected peers.
+    Active,
+}
+
+/// Information about whether the node is connected to other peers and
+/// its consensus status.
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq)]
+pub struct HealthCheckInfo {
+    /// Consensus status
```
Comment (popzxc) [inline exonum/src/api/public/system.rs]: Also dots in some plases are still missing.

File: testkit/tests/system_api.rs [hunk 16]
```diff
@@ -19,20 +19,21 @@ extern crate pretty_assertions;
...
-        private::NodeInfo, public::system::{ConsensusStatusInfo, HealthCheckInfo},
+        private::NodeInfo, public::system::{ConsensusStatus, ConnectivityStatus, HealthCheckInfo},
     },
     helpers::user_agent, messages::PROTOCOL_MAJOR_VERSION,
 };
 use exonum_testkit::{ApiKind, TestKitBuilder};
 
 #[test]
-fn test_healthcheck_connectivity_false() {
+fn test_healthcheck_status_false() {
```
Comment (popzxc) [inline testkit/tests/system_api.rs]: And one more: `test_helthcheck_status_false` name doesn't correlate with test contents.

File: CHANGELOG.md [hunk 17]
```diff
@@ -79,6 +79,8 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
 
 - Added a public `v1/consensus_status` endpoint. (#736)
 
+- Merged public healthcheck and consensus endpoints to `v1/healthcheck`. (#766)
```
Comment (DarkEld3r) [inline CHANGELOG.md]: I suppose this should be a breaking change, right?

Additionally, "Added a public `v1/consensus_status` endpoint. (#736)" should be removed.

**Pull request merged**